kart/dataset_mixins.py

```diff
@@ -46,3 +50,130 @@ def diff_to_wc(self, repo, ds_filter=DatasetKeyFilter.MATCH_ALL):
+        (generally the "same" dataset at a different revision).
+        """
+
+        flags = pygit2.GIT_DIFF_SKIP_BINARY_CHECK
+        self_subtree = self.get_subtree(subtree_name)
+        other_subtree = other.get_subtree(subtree_name) if other else self._empty_tree
+
+        diff = self_subtree.diff_to_tree(other_subtree, flags=flags, swap=reverse)
+        self.L.debug(
+            "diff %s (%s -> %s / %s): %s changes",
+            subtree_name,
+            self_subtree.id,
+            other_subtree.id if other_subtree else None,
+            "R" if reverse else "F",
+            len(diff),
+        )
+        return diff
+
+    _INSERT_UPDATE_DELETE = (
+        pygit2.GIT_DELTA_ADDED,
+        pygit2.GIT_DELTA_MODIFIED,
+        pygit2.GIT_DELTA_DELETED,
+    )
+    _INSERT_UPDATE = (pygit2.GIT_DELTA_ADDED, pygit2.GIT_DELTA_MODIFIED)
+    _UPDATE_DELETE = (pygit2.GIT_DELTA_MODIFIED, pygit2.GIT_DELTA_DELETED)
+
+    def diff_subtree(
+        self,
+        other,
+        subtree_name,
+        key_filter=UserStringKeyFilter.MATCH_ALL,
+        key_decoder_method=None,
+        value_decoder_method=None,
+        reverse=False,
+    ):
+        """
+        Yields deltas from self -> other, but only for items that match the feature_filter.
+        If reverse is true, yields deltas from other -> self.
+        Uses get_raw_diff_from_subtree to find the initial diff, but interprets this diff
+        according to key_decoder_method and value_decoder_method.
+        """
+        # TODO - if the key-filter is very restrictive (ie it has only a few items in) then
+        # it would be more efficient if we first search for those items and diff only those.
+
+        raw_diff = self.get_raw_diff_for_subtree(other, subtree_name, reverse=reverse)
+        # NOTE - we could potentially call diff.find_similar() to detect renames here,
+        #
+
+        if reverse:
+            old, new = other, self
+        else:
+            old, new = self, other
+
+        def get_decoder(dataset, method_name):
+            return (
+                getattr(dataset, method_name)
+                if dataset is not None
+                else lambda key: key
+            )
+
+        old_key_decoder = get_decoder(old, key_decoder_method)
+        new_key_decoder = get_decoder(new, key_decoder_method)
+        old_value_decoder = get_decoder(old, value_decoder_method)
+        new_value_decoder = get_decoder(new, value_decoder_method)
+
+        subtree_path = subtree_name + "/"
+
+        for d in raw_diff.deltas:
+            self.L.debug(
+                "diff(): %s %s %s", d.status_char(), d.old_file.path, d.new_file.path
+            )
+
+            if d.status not in self._INSERT_UPDATE_DELETE:
+                # RENAMED, COPIED, IGNORED, TYPECHANGE, UNMODIFIED, UNREADABLE, UNTRACKED
+                raise NotImplementedError(f"Delta status: {d.status_char()}")
```

Comment:
is this something we need to implement, or could some/all of these delta types simply be ignored? From the comment I expect only RENAMED is something we'll actually care about

Reply:
This has been like this for as long as Sno / Kart has been a thing. You are right that RENAMED is probably the only one that this commit<>commit diff code could potentially handle which it right now doesn't.

Saying the rest "can be ignored" is a bit misleading - right now we don't "ignore" them, we raise an error if we encounter them, and in practise so far we have not encountered them - due to some mixture of the limited things that Kart does and the limited ways we are generating pygit2.Diff objects here (ie commit<>commit only, no rename detection)

I'll add a comment explaining we don't expect to see these here and why.

Reply:
I think if we want other status' we need to pass additional options to libgit2? Which presumably has some performance penalty.

> I'll add a comment explaining we don't expect to see these here and why.

👍